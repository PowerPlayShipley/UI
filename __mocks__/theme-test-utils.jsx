// TODO: Fix this??

import * as React from 'react'
import userEvent from '@testing-library/user-event'
import { render as rtlRender } from '@testing-library/react'

import { ThemeProvider } from "styled-components";

import theme from '../src/themes'
import GlobalStyle from "../src/components/GlobalStyle";

function renderWithTheme(ui, { ...options } = {}) {
  const Wrapper = ({ children }) => (
    <ThemeProvider theme={theme}>
      <GlobalStyle />
      {children}
      <div id='modal-root' />
    </ThemeProvider>
  )

  return rtlRender(ui, { wrapper: Wrapper, ...options })
}

// THE BELLOW COME FROM `https://github.com/styled-components/styled-components/blob/0aa3170c255a49cd41c3fbeb2b8051b5d132f229/src/test/utils.js`

const stripComments = (str) => str.replace(/\/\*.*?\*\/\n?/g, '');

const stripWhitespace = (str) => str
  .trim()
  .replace(/([;\{\}])/g, '$1  ')
  .replace(/\s+/g, ' ');

const getCSS = (scope) =>
  Array.from(scope.querySelectorAll('style'))
    .map(tag => tag.innerHTML)
    .join('\n')
    .replace(/ {/g, '{')
    .replace(/:\s+/g, ':')
    .replace(/:\s+;/g, ':;');

function expectCSSMatches ( _expectation, opts = { ignoreWhitespace: true }) {
  // NOTE: This should normalise both CSS strings to make irrelevant mismatches less likely
  const expectation = _expectation
    .replace(/ {/g, '{')
    .replace(/:\s+/g, ':')
    .replace(/:\s+;/g, ':;');

  const css = getCSS(document);

  if (opts.ignoreWhitespace) {
    const stripped = stripWhitespace(stripComments(css));
    expect(stripped).toEqual(stripWhitespace(expectation));
    return stripped;
  } else {
    expect(css).toEqual(expectation);
    return css;
  }
};

export * from '@testing-library/react';

export { renderWithTheme, expectCSSMatches };
