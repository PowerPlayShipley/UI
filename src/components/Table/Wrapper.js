import styled from "styled-components";

const Wrapper = styled.table`
  border-spacing: 0;
  border: 1px solid ${({ theme }) => theme.main.colors.border.default};
  
  tr {
    :last-child {
      td {
        border-bottom: 0;
      }
    }
  }
  
  td {
    vertical-align: top;
    
    .toolbar {
      display: none;
    }
  }
  
  td:hover .toolbar {
    display: flex;
  }
  
  th, td {
    margin: 0;
    padding: 0.5rem;
    
    border-bottom: 1px solid ${({ theme }) => theme.main.colors.border.default};
    border-right: 1px solid ${({ theme }) => theme.main.colors.border.default};
    
    :last-child {
      border-right: 0;
    }
  }
`

export default Wrapper
