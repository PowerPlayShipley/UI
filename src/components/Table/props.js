import PropTypes from 'prop-types'

const Col = {
  accessor: PropTypes.oneOfType([ PropTypes.string, PropTypes.func ]),
  id: PropTypes.string,
  Header: PropTypes.oneOfType([PropTypes.string, PropTypes.func, PropTypes.element]),
  Footer: PropTypes.oneOfType([PropTypes.string, PropTypes.func, PropTypes.element]),
  Cell: PropTypes.oneOfType([PropTypes.func, PropTypes.element]),
  width: PropTypes.number,
  minWidth: PropTypes.number,
  maxWidth: PropTypes.number
}

export const Column = PropTypes.exact(Col)

export const Columns = PropTypes.arrayOf(PropTypes.exact({
  ...Col,
  accessor: PropTypes.oneOfType([ PropTypes.string, PropTypes.func ]).isRequired,
  columns: PropTypes.arrayOf(Column)
}))
