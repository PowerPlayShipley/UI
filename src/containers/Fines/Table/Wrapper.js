import styled from "styled-components";

const Wrapper = styled.div`
  display: block;
  width: 100%;
  overflow-x: auto;
  
  & table {
    width: 100%;
    max-width: 100%;
    
    .editable div {
      flex-wrap: nowrap;
    }
    
    .editable svg {
      color: ${({ theme }) => theme.main.colors.text.subtle};
    }
  }
  
  & table {
    border: none;
    height: auto;

    border-collapse: separate;
    border-spacing: 0 8px;
    
    th, td {
      width: 1%;
      vertical-align: middle !important;
    }

    th {
      border: none;
      border-right: 1px solid ${({ theme }) => theme.main.colors.border.subtle} !important;
    }

    th:last-child, td:last-child {
      border-right: none !important;
    }
    
    td {
      border: none;
    }
    
    thead {
      font-weight: ${({ theme }) => theme.main.typography.fontWeights.semiBold};
    }
    
    thead {
      th {
        text-align: center;
      }
      
      th:first-child {
        text-align: left;
      }
    }
    
    tbody tr {
      //--scale: 1.01;
      
      background-color: ${({ theme }) => theme.main.colors.system.white};
      box-shadow: ${({ theme }) => theme.main.misc.shadow.medium }, ${({ theme }) => theme.main.misc.inset.default };
      
      //&:hover {
      //  transform: scale(var(--scale));
      //  
      //  & > * {
      //    transform: scale(calc(1 / var(--scale)));
      //  }
      //}
    }
    
    tbody {
      --border-radius-cell: ${({ theme }) => theme.main.misc.radius["2"]};
      
      tr td:first-child { 
        border-top-left-radius: var(--border-radius-cell);
        border-bottom-left-radius: var(--border-radius-cell);
      }
      tr td:last-child {
        border-top-right-radius: var(--border-radius-cell);
        border-bottom-right-radius: var(--border-radius-cell);
      }
    }

    @media (min-width: 767px) {
      // For only large devices
      tbody td {
        .toolbar {
          visibility: hidden;
          opacity: 0;
          transition: visibility 0s, opacity 0.1s linear;
        }

        &:hover {
          .toolbar {
            visibility: visible;
            opacity: 1;
          }
        }
      }
    }
  }
`

export default Wrapper
