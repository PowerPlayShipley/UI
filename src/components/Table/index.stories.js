import React, { useMemo } from 'react'
import { action } from "@storybook/addon-actions";

import Table from "./index";
import EditableCell from "../EditableCell";

export default {
  title: 'Table',
  component: Table
}

const Template = ({ columns, data, ...args }) => {
  const headers = useMemo(() => columns, [columns])
  const body = useMemo(() => data, [data])

  return <Table columns={headers} data={body} {...args} />
};

export const Default = Template.bind({});
Default.args = {
  columns: [
    {
      Header: 'Name',
      accessor: 'name'
    }, {
      Header: 'Age',
      accessor: 'age'
    }
  ],
  data: [{
    name: 'John Smith',
    age: 21
  }, {
    name: 'Jack Smith',
    age: 66
  }]
}

export const EditableDefault = Template.bind({});
EditableDefault.args = {
  toolbarFloat: 'right',
  onToolbarClick: action('toolbarClicked'),
  onItemClick: action('itemClick'),
  columns: [
    {
      Header: 'Name',
      accessor: 'name'
    },
    {
      Header: 'Age',
      accessor: 'age',
      Cell: EditableCell
    },
    {
      Header: 'Another Thing',
      accessor: 'another',
      Cell: EditableCell
    }
  ],
  data: [
    {
      name: 'John Smith',
      age: new Array(21).fill(1),
      another: new Array(16).fill(1)
    },
    {
      name: 'Jack Smith',
      age: new Array(15).fill(1),
      another: new Array(11).fill(1)
    }
  ]
}
