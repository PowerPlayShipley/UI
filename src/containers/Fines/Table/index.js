/**
 * A wrapper for `Table` to handle the fines data.
 *
 * It will take already manipulated fines data and convert them to the headers and the valid data that
 * `react-table` can read.
 * */

import React, { useMemo, memo } from 'react'
import PropTypes from 'prop-types'

import Wrapper from "./Wrapper";

import TBL from "../../../components/Table";
import EditableCell from "../../../components/EditableCell";
import useConfiguration from "../../../hooks/useConfiguration";

/**
 * @typedef {Object} RawData
 *
 * @property {string} player
 * @property {[[number]]} games
 * */

/**
 * @param {RawData[]} data
 * @returns {Object}
 * */
const createHeaders = (data) => {
  let games = 1
  data.forEach((previousValue) => {
    games = Math.max(games, previousValue.games.length)
  })

  let remainingHeaders = new Array(games).fill({
    Cell: EditableCell
  }).map((el, idx) => ({ ...el, Header: `Game ${idx + 1}`, accessor: `game${idx + 1}` }))

  return [{
    Header: 'Player',
    accessor: 'player'
  }, ...remainingHeaders]
}

/**
 * @param {RawData[]} data
 * @param {Object} players
 * @returns {Object}
 * */
const createData = (data, players) => {
  return data.map((el) => ({
    player: el.player in players ? players[el.player].name : el.player,
    ...el.games.reduce(((previousValue, currentValue, currentIndex) => {
      return { ...previousValue, [`game${currentIndex+1}`]: currentValue }
    }), { })
  }))
}

const Table = ({ data: values, onToolbarClick, onItemClick, ...rest }) => {
  const { players } = useConfiguration()

  const columns = useMemo(() => createHeaders(values), [values])
  const data = useMemo(() => createData(values, players), [values, players])

  return (
    <Wrapper {...rest}>
      {data.length > 0 && (<TBL className='table' data={data} columns={columns} onToolbarClick={onToolbarClick} onItemClick={onItemClick} toolbarFloat='right' />)}
    </Wrapper>
  )
}

Table.propTypes = {
  data: PropTypes.arrayOf(
    PropTypes.shape({
      player: PropTypes.string,
      games: PropTypes.array
    })
  ).isRequired,
  onToolbarClick: PropTypes.func,
  onItemClick: PropTypes.func
}

export default memo(Table)
