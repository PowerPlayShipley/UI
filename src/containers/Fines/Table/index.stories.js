import React from 'react'
import { action } from "@storybook/addon-actions";

import Table from "./index";

import ConfigurationProvider from "../../ConfigurationProvider";

export default {
  title: 'Fines/Table',
  component: Table
}

const Template = ({ configuration, data, ...args }) => {
  return (
    <ConfigurationProvider configuration={configuration}>
      <div style={{ backgroundColor: '#f5f5f5', padding: '16px 16px' }}>
        <Table data={data} {...args} />
      </div>
    </ConfigurationProvider>
  )
};

export const Default = Template.bind({});
Default.args = {
  configuration: {
    fines: {},
    players: {
      AA: {
        name: 'Ade'
      },
      TC: {
        name: 'Tom Collins'
      },
      JD: {
        name: 'Jamie'
      }
    }
  },
  data: [
    {
      player: 'AA',
      games: [["H"], ["D", "H", "D", "D"], ["R"]]
    },
    {
      player: 'TC',
      games: [[], ["H"], ["R"]]
    },
    {
      player: 'JD',
      games: [["H", "D", "H", "D", "D"], ["D", "H"], ["R"]]
    }
  ],
  onToolbarClick: action('toolbarClicked'),
  onItemClick: action('itemClick'),
}

// TODO: Add empty UI object
export const Empty = Template.bind({});
Empty.args = {
  configuration: {
    fines: {},
    players: {
      AA: {
        name: 'Ade'
      },
      TC: {
        name: 'Tom Collins'
      },
      JD: {
        name: 'Jamie'
      }
    }
  },
  data: [],
  onToolbarClick: action('toolbarClicked'),
  onItemClick: action('itemClick'),
}
