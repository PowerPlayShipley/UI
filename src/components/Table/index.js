/**
 * Wrapper for `use-table` of `react-table`
 * */

import React, { useEffect } from 'react'
import PropTypes from 'prop-types'
import { useTable } from 'react-table'

import Wrapper from "./Wrapper";
import { Columns, Column } from './props'

/**
 * The rest spread will be added to the table props as anything added in the useTable
 * options will be passed to the cell renderer so we can add out updaters here, keeping
 * this generally generic??
 * */
export default function Table ({ className, columns, data, defaultColumn, onStateChange, plugins = [], ...rest }) {
  const {
    getTableProps,
    getTableBodyProps,
    headerGroups,
    rows,
    prepareRow,
    state
  } = useTable({ columns, data, defaultColumn, ...rest }, ...plugins)

  // Pass the state back to keep this as generic as we can
  useEffect(() => {
    onStateChange && onStateChange(state)
  }, [state, onStateChange])

  return (
    <Wrapper className={className} {...getTableProps()}>
      <thead>
      {headerGroups.map((headerGroup, i) => (
        <tr key={i} {...headerGroup.getHeaderGroupProps()}>
          {headerGroup.headers.map((column, i) => (
            <th key={i} {...column.getHeaderProps()}>
              {column.render('Header')}
            </th>
          ))}
        </tr>
      ))}
      </thead>
      <tbody {...getTableBodyProps()}>
      {rows.map((row, i) => {
        prepareRow(row)
        return (
          <tr key={i} {...row.getRowProps()}>
            {row.cells.map((cell, i) => {
              return <td key={i} {...cell.getCellProps()}>{cell.render('Cell')}</td>
            })}
          </tr>
        )
      })}
      </tbody>
    </Wrapper>
  )
}

Table.propTypes = {
  className: PropTypes.string,
  columns: Columns.isRequired,
  data: PropTypes.array.isRequired,
  onStateChange: PropTypes.func,
  defaultColumn: Column,
  // Use any react-table plugins
  plugins: PropTypes.arrayOf(PropTypes.func)
}
