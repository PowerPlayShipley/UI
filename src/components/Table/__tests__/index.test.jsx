/**
 * @jest-environment jsdom
 */

import React from 'react'
import { renderWithTheme } from '../../../../__mocks__/theme-test-utils'

import Table from "../index";

describe('Display Table', () => {
  it('should create valid snapshot', () => {
    const cols = [
      {
        Header: 'Name',
        accessor: 'name'
      }, {
        Header: 'Age',
        accessor: 'age'
      }
    ]

    const data = [{
      name: 'John Smith',
      age: 21
    }, {
      name: 'Jack Smith',
      age: 66
    }]

    const { container } = renderWithTheme(<Table data={data} columns={cols} />)
    expect(container.firstChild).toMatchSnapshot()
  })
})
