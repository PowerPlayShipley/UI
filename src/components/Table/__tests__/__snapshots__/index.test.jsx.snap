// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Display Table should create valid snapshot 1`] = `
.c0 {
  border-spacing: 0;
  border: 1px solid #d0d7de;
}

.c0 tr:last-child td {
  border-bottom: 0;
}

.c0 td {
  vertical-align: top;
}

.c0 td .toolbar {
  display: none;
}

.c0 td:hover .toolbar {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c0 th,
.c0 td {
  margin: 0;
  padding: 0.5rem;
  border-bottom: 1px solid #d0d7de;
  border-right: 1px solid #d0d7de;
}

.c0 th:last-child,
.c0 td:last-child {
  border-right: 0;
}

<table
  class="c0"
  role="table"
>
  <thead>
    <tr
      role="row"
    >
      <th
        colspan="1"
        role="columnheader"
      >
        Name
      </th>
      <th
        colspan="1"
        role="columnheader"
      >
        Age
      </th>
    </tr>
  </thead>
  <tbody
    role="rowgroup"
  >
    <tr
      role="row"
    >
      <td
        role="cell"
      >
        John Smith
      </td>
      <td
        role="cell"
      >
        21
      </td>
    </tr>
    <tr
      role="row"
    >
      <td
        role="cell"
      >
        Jack Smith
      </td>
      <td
        role="cell"
      >
        66
      </td>
    </tr>
  </tbody>
</table>
`;
